### PR TITLE
Safe columns

### PR DESCRIPTION
**Pull request by peterhurford**
Many times when new columns are discovered in the cache, this is not a good thing.  Instead, it is indicative of cache invalidation issues!  Now we can safeguard against that with `safe_columns = TRUE`.

If `safe_columns = TRUE` on your caching layer and if a call would add more columns while columns already exist, it stops with an error.  If you want to cache it, you'll have to drop the entire cache and start fresh (which should be happening if column additions are happening).

If `safe_columns` is a function, that custom function will be called, so you can do interactive debugging.

### CODE REVIEW

**Comment by peterhurford** [comment]
^ Not a real failure

File: R/db.R [hunk 1]
```diff
@@ -168,7 +169,16 @@ write_data_safely <- function(dbconn, tblname, df, key) {
...
       column_map <- column_map[!is.element(column_map$raw_name, raw_names), ]
       if (NROW(column_map) > 0) {
-        dbWriteTable(dbconn, 'column_names', column_map, append = TRUE, row.names = FALSE)
+        if (isTRUE(safe_columns)) {
+          stop("Safe Columns Error: Your function call is adding additional ",
+            "columns to a cache that already has pre-existing columns. This ",
+            "would suggest your cache is invalid and you should wipe the cache ",
+            "and start over.")
+        } else if (is.function(safe_columns)) {
+          safe_columns()
```
Comment (robertzk) [inline R/db.R:178]: This is kind of weird. Should we be doing line 180 as well right after this call?

File: R/db.R [hunk 2]
```diff
@@ -168,7 +169,16 @@ write_data_safely <- function(dbconn, tblname, df, key) {
       raw_names <- DBI::dbGetQuery(dbconn, 'SELECT raw_name FROM column_names')[[1]]
       column_map <- column_map[!is.element(column_map$raw_name, raw_names), ]
       if (NROW(column_map) > 0) {
-        dbWriteTable(dbconn, 'column_names', column_map, append = TRUE, row.names = FALSE)
+        if (isTRUE(safe_columns)) {
+          stop("Safe Columns Error: Your function call is adding additional ",
+            "columns to a cache that already has pre-existing columns. This ",
+            "would suggest your cache is invalid and you should wipe the cache ",
+            "and start over.")
```
Comment (robertzk) [inline R/db.R:176]: Can you print what columns in the error message and save them to a global variable?

File: README.md [hunk 3]
```diff
@@ -256,6 +259,24 @@ Shard dimensions:
...
 
+## safe_columns
+
+If you have trouble with your cache ending up writing additional columns when
+this isn't supposed to be happening (e.g., an event that should be triggering
+cache invalidation), you can use `safe_columns = TRUE` as a parameter to `cache`
+in order to error instead of writing additional columns.
+
+Additionally, you can pass a function as `safe_columns = fn` as long as that
+function takes no arguments and returns `TRUE`.  This function will then be called
```
Comment (robertzk) [inline README.md:270]: Nice!

File: R/db.R (hunk 1 above)
Comment (robertzk) [inline R/db.R:178]: I see -- nevermind, nice!

**Comment by robertzk** [comment]
Thanks for the test coverage.

**Pull request merged**